# <center> Profitable App Profile for the Apple Store and Google Play Markets </center>

<center> by: Chlaudiah Julinar Soplero Lelywiary </center>

This project is to find the characteristics of mobile appplication that are prefreable in Apple Store and Google Play Store. As a data analyst who work in company that builds Android and iOS mobile apps, the job is to help the developers team to make a decision based on data-driven approach for decided what kind of apps that they should build. 

The company only build apps that are free to download and install and main source of revenue consites of in-app adds. This means our revenue for any given app is mostly influenced by the number of users who use our app. Our goal for this project is to analyze data to help our developers understand what kind of apps are likely to attract more users.

## Opening and Exploring the Data

As of September 2018, there were approximately 2 million iOS apps available on the App Store and 2.1 million Android apps on Google Play (source from: [Statista](https://www.statista.com/statistics/276623/number-of-apps-available-in-leading-app-stores/))

Collectiong data over 4 million apps requires a significant amount of time and money, so try to analyze a sample of the data instead. To avoid spending resources on collectiong new data ourselves try to see any relevant existing data without cost. There are two datasets below seem suitable for the goals:

* [Google Apps Store Datasets](https://www.kaggle.com/lava18/google-play-store-apps/home) containing data about approximately **10.000** Android apps from Google Play that collected in August 2018.
* [Apple iOS Apps Store Datasets](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home) containing data about approximately **7.000** iOS apps from the App Store that collected in July 2017.

Lets start by opening two data sets and then continue with exploring the data.

In [1]:
## OPEN TWO DATA SETS#

from csv import reader

#------- Google App Store Data sets------#
open_android = open('googleplaystore.csv',encoding='utf-8')
android_file = reader(open_android)
android_lists = list(android_file)
android_header = android_lists[0]
android_data = android_lists[1:]

#------- Apple iOS App Store Data sets------#
open_ios = open('AppleStore.csv',encoding='utf-8')
ios_file = reader(open_ios)
ios_lists = list(ios_file)
ios_header = ios_lists[0]
ios_data = ios_lists[1:]

In [18]:
# EXPLORE TWO DATA SETS #
def explore_data(lists, header=False):
    for row in lists[:3]:
        print(row,'\n')
        
    if header:
        print('Number of rows: ',len(lists))
        print('Number of columns: ',len(lists[0]))

In [3]:
print('Android Data Explore: ')
android_explore = explore_data(android_lists, header=True)

Android Data Explore: 
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up'] 

Number of rows:  10841
Number of columns:  13


As we can see from the result, for Android Dataset has **13 feature** and **10841 data**. The features that might be useful for our goals are:

| <center> Column Name </center> | <center> Description </center>                                |
| ----------- | ------------------------------------------- |
| App         | Application Name                            |
| Category    | Category the app belongs to                 |
| Rating      | Overall user rating of the app              |
| Reviews     | Number of user reviews                      |
| Installs    | Number of user download or install the app  |
| Type        | a paid app or free app                      |
| Genres      | Apart from the main category                |
| Price       | Price of the app                            |

In [4]:
print('iOS Data Explore: ')
ios_explore = explore_data(ios_lists,True)

iOS Data Explore: 
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'] 

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'] 

Number of rows:  7197
Number of columns:  16


For iOS data sets has **16 features** and **7197 data**. The features that might be useful are:

| Column Name       |  Description                             |
| -----------       | -----------------------------------------|
| id                | Application ID                           |
| track_name        | Application Name                         |
| price             | price amount                             |
| rating_count_tot  | Number of user rating count for all ver  |
| rating_count_ver  | Number of user rating count for cur ver  |
| prime_genre       | primary genre                            |
| currency          | Currency Type                            |

## Data Cleaning

### Detecting and Deleting Wrong Data

The Google Play data set has a dedicated [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion), and [one of the discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) describes an error for a certain row.



In [5]:
print('Wrong data: ',android_data[10472],'\n')
print('List of Features: ',android_header,'\n')
print('Example of right data: ',android_data[0])

Wrong data:  ['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up'] 

List of Features:  ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

Example of right data:  ['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


It is right from the discussion, that the data in index:**10472** (without header) is a wrong data because **Content Rating** feature has a blank value. Also, as we can see the **Rating** of that app reach **19**, and it is clearly off because the maximum rating for a Google Play App is a **5**.

So, need to delete the row of that wrong data from the Google Play data set.

In [6]:
del android_data[10472]
print(len(android_data))

10840


### Removing Duplicate Entries

#### Part One

After cleaning the wrong data, in this part will delete the duplicate entries in Google Play data set (look at the [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion)). Duplicate entries cases where an application appears more than once in a data set. For example, Instagram has four entries in Google Play data set:

In [7]:
for i in range(len(android_data)):
    name=android_data[i][0]
    if name=='Instagram':
        print(i,android_data[i])

2545 ['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
2604 ['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
2611 ['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
3909 ['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


We will count how many cases are duplicate entries in Google Play data set.

In [8]:
duplicate = []
unique=[]
for row in android_data:
    name=row[0]
    if name in unique:
        duplicate.append(name)
    else: unique.append(name)
print("Number of duplicate entries: ",len(duplicate))
print("Example of duplicate entries: ",duplicate[:5])

Number of duplicate entries:  1181
Example of duplicate entries:  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


From the result, we can see that there are 1,181 cases for duplicate entries in Google Play datset. To handle this case, we don't want to count certain apps more than once when analyze data, so we need to remove the duplicate entries and keep only one entry per app. One thing we could do is remove the duplicate rows randomly, but there is still another way and better than this. 

If we back to Instagram's rows example, there are four difference in every row, and that is in index 3 (remember that list in Python start with 0). To find what column is it for index 3, then we can print it below:

In [9]:
for i in range(len(android_header)):
    print(i,android_header[i])

0 App
1 Category
2 Rating
3 Reviews
4 Size
5 Installs
6 Type
7 Price
8 Content Rating
9 Genres
10 Last Updated
11 Current Ver
12 Android Ver


So, the differences that occurs in every Instagram's row is number of reviews. The different numbers show the data was collected at different times. Using this information, the criterion for removing the duplicates entries can be build. Main important thing that we can get from this information is **the higher number of reviews, the more recent the data should be**.

#### Part Two

Instead of removing the duplicates randomly, just keep the row with the highest number of reviews and remove the other entries for any given app. To remove the duplicates, the steps are:
1. Create a dictionary, where each dictionary key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app
2. Use the information stored in dictionary and create a new data set, which will have only entry per app (and for each app, only select the entry with the highest number of reviews)

##### Step 1

In this step, we build a dictionary to save the highest number of review from every app. 

In [10]:
reviews_max={}
for row in android_data:
    name = row[0]
    n_reviews=float(row[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name]=n_reviews
    reviews_max[name]=n_reviews

In **Part One** we knew that there are **1,181 of duplicate entries** that we have. So, for make sure our dictionary doesnt have an app occurs more than once (of unique apps) it should be equal to the difference between the length of our data set and 1,181.

In [11]:
print("Expected Length: ",len(android_data)-1181)
print("Actual Length of Dictionary: ",len(reviews_max))
print("Example of the Highest number of review for Instagram: ",reviews_max["Instagram"])

Expected Length:  9659
Actual Length of Dictionary:  9659
Example of the Highest number of review for Instagram:  66509917.0


##### Step 2

Next, we need to make a new data set that stored only one entry per app with the highest number of reviews. In the code below:

- Create two empty lists, `android_clean` (will store new data set) and `already_dead` (will store app names)
- Do looping in Google Play data set without header (`android_data`) for each row:
    - Assign app name (in index 0)
    - Convert the type of number of reviews (in index 3) to `float`  and asssign it into a variable
    - Add the current row into `android_clean` and the app name into `already_dead`, **if**:
        - The number of reviews equal with the highest number of reviews from `reviews_max` dictionary for that app; and
        - The name of the app not already in `already_dead`. This part is important to handle a case when there are more than one entry has highest number of reviews for an app.


In [16]:
android_clean = []
already_dead=[]
for row in android_data:
    name = row[0]
    n_reviews = float(row[3])
    if (n_reviews == reviews_max[name]) and (name not in already_dead):
        android_clean.append(row)
        already_dead.append(name)

Now, just make sure that our new data set already same with the criterion using `explore_data` function that we already build and the number of rows is **9659**.

In [19]:
explore_data(android_clean,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'] 

Number of rows:  9659
Number of columns:  13


### Removing Non-English Apps

#### Part One